portal-impl/src/com/liferay/portal/db/remover/PortalDuplicateRemover.java

```diff
@@ -0,0 +1,212 @@
+package com.liferay.portal.db.remover;
+
+import com.liferay.portal.kernel.dao.jdbc.DataAccess;
+import com.liferay.portal.kernel.util.StringBundler;
+import com.liferay.portal.kernel.util.StringUtil;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.ResultSetMetaData;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+public class PortalDuplicateRemover implements DuplicateRemover {
+
+	@Override
+	public Map<Long, List<HashMap<String, String>>> getDuplicates(
+		String tableName, String indexesSQL) {
+
+		Map<String[], String> indexesDuplicatesMap = new LinkedHashMap<>();
+
+		Map<Long, List<HashMap<String, String>>> duplicatesMap =
+			new LinkedHashMap<>();
+
+		List<String> indexesColumnsList = getIndexesColumnsList(indexesSQL);
+
+		for (String indexColumns : indexesColumnsList) {
+			StringBundler sb = new StringBundler(7);
+
+			sb.append("SELECT ");
+			sb.append(indexColumns);
+			sb.append(" FROM");
+			sb.append(tableName);
+			sb.append(" GROUP BY ");
+			sb.append(indexColumns);
+			sb.append(" HAVING COUNT(*) > 1;");
+			String sql = sb.toString();
+
+			try (Connection connection = DataAccess.getConnection();
+				PreparedStatement preparedStatement1 =
+					connection.prepareStatement(sql);
+				ResultSet resultSet = preparedStatement1.executeQuery()) {
+
+				ResultSetMetaData metaData = resultSet.getMetaData();
+
+				int columnCount = metaData.getColumnCount();
+
+				String[] columnResults = new String[columnCount];
+
+				while (resultSet.next()) {
+					for (int i = 1; i <= columnCount; i++) {
+						String value = resultSet.getString(i);
+
+						columnResults[i - 1] = value;
+					}
+
+					indexesDuplicatesMap.put(columnResults, indexColumns);
+				}
+			}
+			catch (Exception e) {
+				throw new RuntimeException(e);
+			}
+		}
+
+		Long duplicateGroupNumber = 1L;
+
+		for (Map.Entry<String[], String> entry :
+				indexesDuplicatesMap.entrySet()) {
+
+			String[] results = entry.getKey();
+
+			String[] columns = entry.getValue(
+			).split(
+				", "
+			);
+
+			StringBundler sb = new StringBundler();
+
+			sb.append("SELECT * FROM");
```

Comment:
This second query can be improved to use the timestamp column if available to delete the most recent entry,  however this assumes that the entry is valid.

Reply:
Please see mu comment above

Reply:
If needed, to get the most recent entry you can just sort by the ID column (except for Company table since generates a random ID)